deltacat/storage/rivulet/schema/schema.py

```diff
@@ -1,207 +1,261 @@
+        #    raise TypeError("The first field set on a Schema must be a merge key.")
+
+        self._fields[processed_field.name] = processed_field
 
     def __delitem__(self, key: str):
-        if key == self.primary_key.name:
-            raise ValueError("Cannot delete the primary key field")
-        del self.fields[key]
+        field = self._fields[key]
+        if field.is_merge_key:
+            raise ValueError("Cannot delete a merge key field")
+        del self._fields[key]
 
-    def __repr__(self):
-        field_reprs = [
-            f"{name}: {field.datatype}" for name, field in self.fields.items()
-        ]
-        return (
-            f"Schema({', '.join(field_reprs)}, primary_key='{self.primary_key.name}')"
-        )
+    def __len__(self):
+        return len(self._fields)
 
     def __iter__(self):
-        return iter(self.fields.values())
+        return iter(self._fields.keys())
+
+    def __hash__(self):
+        return hash((frozenset(self._fields.items())))
+
+    def __eq__(self, other):
+        if isinstance(other, Schema):
+            return self._fields == other._fields
+        return False
+
+    # Has a spurious type check problem in @dataclass + asdict(): https://youtrack.jetbrains.com/issue/PY-76059/Incorrect-Type-warning-with-asdict-and-Dataclass
+    def to_dict(self):
+        return {"fields": [asdict(field) for field in self._fields.values()]}
+
+    def add_field(self, field: Field):
+        """Adds a Field object using its name as the key"""
+        self[field.name] = field
+
+    def get_merge_keys(self) -> Iterable[str]:
+        """Return a list of all merge keys."""
+        return [field.name for field in self._fields.values() if field.is_merge_key]
+
+    def get_merge_key(self) -> str:
+        """Returns a single merge key if there is one, or raises if not. Used for simple schemas w/ a single key"""
+        # Get the merge key
+        merge_keys = list(self.get_merge_keys())
+        if len(merge_keys) != 1:
+            raise ValueError(
+                f"Schema must have exactly one merge key, but found {merge_keys}"
+            )
+        return merge_keys[0]
+
+    def merge(self, other: Schema) -> None:
+        """Merges another schema's fields into the current schema."""
+        if not other:
+            return
+        for name, field in other._fields.items():
+            if name in self._fields:
+                if self._fields[name] != field:
+                    raise ValueError(
+                        f"Field '{name}' already exists in the current schema with different definition"
+                    )
+            else:
+                self.add_field(field)
+
+    def to_pyarrow(self) -> pa.Schema:
+        """
+        Convert the Schema to a PyArrow schema.
+
+        Returns:
+            pyarrow.schema: A PyArrow schema representation of this Schema.
+        """
+        # TODO: Should we track merge_keys as it goes to/from pyarrow?
```

Comment:
My opinion: yes. You can see how merge keys and other similar schema annotations are already tracked via https://github.com/ray-project/deltacat/blob/2.0/deltacat/storage/model/schema.py#L345 and https://github.com/ray-project/deltacat/blob/2.0/deltacat/storage/model/schema.py#L252. This metadata is also preserved to/from Parquet. Note that `field_id` at https://github.com/ray-project/deltacat/blob/2.0/deltacat/storage/model/schema.py#L18 is an existing metadata field convention already used by Parquet and Iceberg.